test/spmd/tensor/test_megatron_example.py

```diff
@@ -58,23 +41,56 @@ def shard_module(m, device_type):
     col_wise_sharding = [Shard(0)]
     row_wise_sharding = [Shard(1)]
     replicate = [Replicate()]
-    m.net1.weight = torch.nn.Parameter(
-        distribute_tensor(m.net1.weight, device_mesh, col_wise_sharding)
-    )
-    m.net2.weight = torch.nn.Parameter(
-        distribute_tensor(m.net2.weight, device_mesh, row_wise_sharding)
-    )
-    m.net1.bias = torch.nn.Parameter(
-        distribute_tensor(m.net1.bias, device_mesh, col_wise_sharding)
-    )
-    m.net2.bias = torch.nn.Parameter(
-        distribute_tensor(m.net2.bias, device_mesh, replicate)
-    )
-    m = _replicate_input_tensor(m, device_mesh, replicate)
-    m.net2 = _aggregate_local_tensor(m.net2)
-    m.net1.weight.register_hook(
-        functools.partial(_gradient_hook, m.net1.weight)
+
+    def shard_params(name, module):
+        if isinstance(module, nn.Linear):
+            if name == "net1":
```

Comment:
Hmm, this is kind of hacky, right? Can we have something like a spec or json input as input like shard_module? I mean basically, can we make this shard_params a util function available for all users not just an internal api inside the test?

Reply:
I don't think `shard_module` api provide the flexibility of doing anything we want, for example the gradient hook you need to install for all module params, there might be other things advanced user would like to control, providing a partition_fn could make it more flexible, if we make it a spec or json, it's tight to each module architecture and user need to modify it whenever the module architecture changes. To provide the ease of use, we should provide a bunch of default sharding strategies (pre-set sharding funcs) to let user directly use them instead of writing their own. Does this make sense to you?